src/main/java/nl/ordina/jobcrawler/service/VacancyService.java

```diff
@@ -1,58 +1,92 @@
-import java.util.stream.Collectors;
+import java.util.*;
 
 @Service
 public class VacancyService {
 
-    private VacancyRepository repository;
+    private VacancyRepository vacancyRepository;
+    private SkillService skillService;
 
     @Autowired
-    public VacancyService(VacancyRepository repository) {
-        this.repository = repository;
+    public VacancyService(VacancyRepository vacancyRepository, SkillService skillService) {
+        this.vacancyRepository = vacancyRepository;
+        this.skillService = skillService;
     }
 
-    /* Add */
+
+    //******** Adding ********//
     public Vacancy add(Vacancy vacancy) {
-        return repository.save(vacancy);
-    }
 
-    /* Get */
-    public List<Vacancy> getAllJobs() {
-        return repository.findAll();
+        Set<Skill> ExistingSkills = skillService.linkToExistingSkills(vacancy.getSkills(), vacancy);
+        vacancy.setSkills(ExistingSkills);
+
+        if (vacancy.checkURL()) return vacancyRepository.saveAndFlush(vacancy);
+        else return null;
     }
 
-    public List<Vacancy> getJobsWithSkill(String skill) {
-        return repository.findAll()
-                .stream()
-                .filter(a -> a.getSkillSet().toString().toLowerCase().contains(skill.toLowerCase()))
-                .collect(Collectors.toList());
+    //******** Getting ********//
+    public List<Vacancy> getAllJobs() {
+        return vacancyRepository.findAll();
     }
 
     public Optional<Vacancy> getByID(UUID id) {
-        return repository.findById(id);
+        return vacancyRepository.findById(id);
     }
 
-    public Optional<Vacancy> getExistingRecord(String url) {
-        return repository.findByVacancyURLEquals(url);
+    public Set<Vacancy> getJobsWithSkill(String skill) {
+        return skillService.getVacanciesBySkill(skill);
     }
 
     public List<Vacancy> getJobsByBroker(String broker) {
-        return repository.findByBrokerEquals(broker);
+        return vacancyRepository.findByBrokerEquals(broker);
     }
 
-    /* Delete */
+    public Optional<Vacancy> getExistingRecord(String url) {
+        return vacancyRepository.findByVacancyURLEquals(url);
+    }
+
+
+    //******** Deleting ********//
     public void delete(UUID id) {
-        repository.deleteById(id);
+        Vacancy vacancyToDelete = vacancyRepository.findById(id).orElseThrow(() -> new VacancyNotFoundException(id));
+        Set<Skill> skillsToDelete = new HashSet<Skill>(vacancyToDelete.getSkills());
+
+        vacancyRepository.deleteById(id);
+        System.out.println(skillsToDelete);
+        skillService.deleteSkillsIfNoRelations(skillsToDelete);
+
+    }
+
+
+    //******** Updating ********//
+    public Vacancy replace(UUID id, Vacancy newJob) {
+        return vacancyRepository.findById(id)
+                .map(job -> {
+                    System.out.println("** Job before modification: \n" + job);
+
+                    job.setVacancyURL(newJob.getVacancyURL());
```

Comment:
where does the repo skill check if things are the same

Reply:
It sends an update query for the first time it is called, then if there are no changes it does not send a query, unless there is a change. 
For this function there is a Jira sub-task (NEX-26) to have all the attributes optional when updating a vacancy